includes/server/class-llms-rest-memberships-controller.php

```diff
@@ -0,0 +1,590 @@
+
+		// Sales page id.
+		if ( 'page' === $data['sales_page_type'] || 'edit' === $context ) {
+			$data['sales_page_page_id'] = $membership->get( 'sales_page_content_page_id' );
+		}
+
+		// Sales page url.
+		if ( 'custom' === $data['sales_page_type'] || 'edit' === $context ) {
+			$data['sales_page_url'] = $membership->get( 'sales_page_content_url' );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Register routes.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		parent::register_routes();
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<id>[\d]+)/enrollments',
+			array(
+				'args'   => array(
+					'id' => array(
+						'description' => __( 'Unique Membership Identifier. The WordPress Post ID', 'lifterlms' ),
+						'type'        => 'integer',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this->enrollments_controller, 'get_items' ),
+					'permission_callback' => array( $this->enrollments_controller, 'get_items_permissions_check' ),
+					'args'                => $this->enrollments_controller->get_collection_params(),
+				),
+				'schema' => array( $this->enrollments_controller, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Updates a single llms membership.
+	 *
+	 * @since [version]
+	 *
+	 * @param LLMS_Membership $membership    LLMS_Membership instance.
+	 * @param WP_REST_Request $request       Full details about the request.
+	 * @param array           $schema        The item schema.
+	 * @param array           $prepared_item Array.
+	 * @param bool            $creating      Optional. Whether we're in creation or update phase. Default true (create).
+	 * @return bool|WP_Error True on success or false if nothing to update, WP_Error object if something went wrong during the update.
+	 */
+	protected function update_additional_object_fields( $membership, $request, $schema, $prepared_item, $creating = true ) {
+		$error = new WP_Error();
+
+		// Auto enroll.
+		if ( ! empty( $schema['properties']['auto_enroll'] ) && isset( $request['auto_enroll'] ) ) {
+			$membership->add_auto_enroll_courses( $request['auto_enroll'], true );
+		}
+
+		// Catalog visibility.
+		if ( ! empty( $schema['properties']['catalog_visibility'] ) && isset( $request['catalog_visibility'] ) ) {
+			$membership->get_product()->set_catalog_visibility( $request['catalog_visibility'] );
+		}
+
+		// Instructors.
+		if ( ! empty( $schema['properties']['instructors'] ) && isset( $request['instructors'] ) ) {
+
+			$instructors = array();
+
+			foreach ( $request['instructors'] as $instructor_id ) {
+				$user_data = get_userdata( $instructor_id );
+				if ( ! empty( $user_data ) ) {
+					$instructors[] = array(
+						'id'   => $instructor_id,
+						'name' => $user_data->display_name,
+					);
+				}
+			}
+
+			$membership->set_instructors( $instructors );
```

Comment:
I think this misses the case of creating a membership without instructors.
Basically we always want at least an instructor to be set.

I guess you took this code from the course resources, which misses this case too.
I've amended it in a pending PR #98, which, anyways while reviewing your code I realized is still not perfect, and I've just updated it, and updated the membership code as well, see:
https://github.com/eri-trabiccolo/lifterlms-rest/blob/memberships/includes/server/class-llms-rest-memberships-controller.php#L446

Do you find anything wrong with my code? Let me know.

Reply:
If a membership is created or saved without any instructors, `LLMS_Post_Instructors->pre_set_instructors()` will default to the current user. See my question issue #129 about what should happen with bad instructors or empty instructors. I am adding tests for these scenarios. If the `llms_validate_instructors()` function in PR #98 is used and instructors are empty or contain invalid instructors, `LLMS_REST_Memberships_Controller->update_additional_object_fields()` is not executed.

Reply:
Sorry, I'm not in front at the computer right now, so I might be wrong but... I'm not sure what you mean with this:
> If a membership is created or saved without any instructors, LLMS_Post_Instructors->pre_set_instructors() will default to the current user.

"Who" calls the `pre_set_instructors()` (`set_instructors()`) on membership creation if no `instructors` is provided in the request?!

Reply:
In the first version of this PR, `set_instructors()` is called by `LLMS_REST_Memberships_Controller->update_additional_object_fields()` if `! empty( $schema['properties']['instructors'] ) && isset( $request['instructors'] )`. If `$request['instructors']` is set, the id's are validated.

Reply:
That's what I meant. If no instructors are provided on creation `isset( $request['instructors'])` is false hence `set_instructors()` is not executed... Am I missing anything?

Reply:
Ah! I get it now. What happens when `instructors` are not given in the request?
1. `LLMS_REST_Memberships_Controller->create_llms_post()` creates a new `LLMS_Membership` object.
2. `LLMS_Post_Model->get_creation_args()' sets `post_author` to `get_current_user_id()`.
3. Later, `LLMS_Post_Instructors->get_instructors()` defaults to `post_author`.

So you are saying we should make certain that the membership has it's instructors set and not let instructors be empty and have `LLMS_Post_Instructors->get_instructors()` make it _appear_ to be set to `post_author`. I will make it so.

Reply:
Yeah, which basically reduces to always call `set_instructors($instructors)` on creation, whit instructors==empty array on creation if no instructors are provided. :)